File: game304/dd.py
"""Per-world double-dummy solver for caps obligation.

Implements the inner minimax of §5 of ``docs/caps_formalism.md``:
given a fixed :class:`game304.info.World` (a concrete card location
for every seat) and a fixed caller play order, decide whether the
order wins every remaining round against any legal continuation by
the other three seats.

The outer single-dummy quantifiers — universal over consistent worlds
and existential over orders — live in :mod:`game304.caps`. This
module is a leaf: it knows nothing about information sets.

Two predicates are exposed:

- :func:`order_sweeps_world` — caps test. The caller's order must
  win every remaining round against the worst-case adversary play.
- :func:`order_min_points_in_world` — claim-balance test.  Returns
  the minimum points the caller's team is guaranteed to collect from
  the remaining play, against the worst-case adversary play. The
  caller's order is fixed; the other three seats minimise.

Adversary legality covers must-follow-suit and the exhausted-trumps
rule (the leader holding all remaining trumps must lead trump). The
closed-trump face-down rules and the in-hand-trump fold restriction
are not enforced in the witness because the announced caps order is
played face-up; once the caller exposes their hand, fold semantics
are moot.
"""

from __future__ import annotations

from dataclasses import dataclass

from game304.card import Card
from game304.info import World
from game304.seating import next_seat, team_of
from game304.types import Seat, Suit, Team


@dataclass(frozen=True)
class InProgressEntry:
    """A play already made in the in-progress round, identity resolved.

    Identities of face-down entries hidden from the viewer must be
    materialised by the caller using the world's hidden-slot
    assignments before being passed to the solver.
    """

    seat: Seat
    card: Card


@dataclass(frozen=True)
class PlaySnapshot:
    """The state of the in-progress round at the moment of analysis.

    Attributes:
        leader: The seat that leads the in-progress round (the seat
            with priority).
        entries: Plays already made in the in-progress round, in play
            order, with face-down identities resolved per the world.
    """

    leader: Seat
    entries: tuple[InProgressEntry, ...]


# ---------------------------------------------------------------------------
# Public predicates
# ---------------------------------------------------------------------------


def order_sweeps_world(
    *,
    world: World,
    caller_seat: Seat,
    caller_order: list[Card],
    snapshot: PlaySnapshot,
    pcc_partner_out: Seat | None,
    rounds_remaining: int,
) -> bool:
    """Return ``True`` iff ``caller_order`` wins every remaining round.

    Adversaries (the other three seats) play minimax-optimally to
    break the sweep. Forced plays (one-card-of-led-suit, exhausted
    trumps that force a trump lead) collapse the adversarial branch
    automatically — that is the deducible-via-partner machinery from
    §5.

    The caller's order must contain exactly the caller's remaining
    cards. ``rounds_remaining`` is the number of rounds left to play
    (including the in-progress one if it has plays).
    """
    sim_hands = {seat: list(cards) for seat, cards in world.hands.items()}
    in_progress = [(e.seat, e.card) for e in snapshot.entries]
    return _solve_caps(
        sim_hands=sim_hands,
        caller_seat=caller_seat,
        caller_order=caller_order,
        caller_index=0,
        leader=snapshot.leader,
        in_progress=in_progress,
        rounds_remaining=rounds_remaining,
        trump_suit=world.trump_suit,
        my_team=team_of(caller_seat),
        pcc_partner_out=pcc_partner_out,
    )


def order_min_points_in_world(
    *,
    world: World,
    caller_seat: Seat,
    caller_order: list[Card],
    snapshot: PlaySnapshot,
    pcc_partner_out: Seat | None,
    rounds_remaining: int,
) -> int:
    """Return the minimum points the caller's team is guaranteed to win.

    Computed against worst-case adversary play. The caller's order is
    fixed; the other three seats play any legal continuation, and the
    return value is the minimum total points scored by the caller's
    team across the remaining rounds.

    Used by Claim Balance: the predicate "we will reach our threshold"
    is equivalent to ``points_already_won + min_points >= threshold``.
    """
    sim_hands = {seat: list(cards) for seat, cards in world.hands.items()}
    in_progress = [(e.seat, e.card) for e in snapshot.entries]
    return _solve_min_points(
        sim_hands=sim_hands,
        caller_seat=caller_seat,
        caller_order=caller_order,
        caller_index=0,
        leader=snapshot.leader,
        in_progress=in_progress,
        rounds_remaining=rounds_remaining,
        trump_suit=world.trump_suit,
        my_team=team_of(caller_seat),
        pcc_partner_out=pcc_partner_out,
    )


# ---------------------------------------------------------------------------
# Internals — round mechanics shared by both solvers
# ---------------------------------------------------------------------------


def _round_turn_order(
    leader: Seat, pcc_partner_out: Seat | None
) -> list[Seat]:
    """Anticlockwise turn order for one round, skipping the PCC-out seat."""
    target = 3 if pcc_partner_out is not None else 4
    order: list[Seat] = [leader]
    cur = leader
    while len(order) < target:
        cur = next_seat(cur)
        if cur == leader:
            break
        if cur == pcc_partner_out:
            continue
        order.append(cur)
    return order


def _round_winner(
    plays: list[tuple[Seat, Card]], trump_suit: Suit | None
) -> Seat:
    """Return the winning seat of a complete round."""
    if not plays:
        raise ValueError("No plays to resolve")
    led_suit = plays[0][1].suit
    trumps = [
        (s, c) for s, c in plays
        if trump_suit is not None and c.suit == trump_suit
    ]
    if trumps:
        return min(trumps, key=lambda sc: sc[1].power)[0]
    led = [(s, c) for s, c in plays if c.suit == led_suit]
    if not led:
        return plays[0][0]
    return min(led, key=lambda sc: sc[1].power)[0]


def _round_points(plays: list[tuple[Seat, Card]]) -> int:
    """Sum of point values for a complete round."""
    return sum(c.points for _, c in plays)


def _legal_plays(
    *,
    hand: list[Card],
    led_suit: Suit | None,
    trump_suit: Suit | None,
    is_lead: bool,
    seats_with_trumps: set[Seat],
    seat: Seat,
) -> list[Card]:
    """Cards this ``seat`` can legally play given the round state.

    Enforces:

    - **Follow suit**: when not leading and holding the led suit,
      only led-suit cards are legal.
    - **Exhausted trumps**: when leading and this seat is the only
      seat holding any remaining trump, the seat must lead trump (if
      they have any) until their own trumps are gone.

    Closed-trump face-down semantics are intentionally absent: caps
    cards are exposed face-up, so the fold rules do not apply to a
    declared play order.
    """
    if not is_lead:
        suited = [c for c in hand if c.suit == led_suit]
        return suited if suited else list(hand)
    # Leading.
    if (
        trump_suit is not None
        and seats_with_trumps == {seat}
        and any(c.suit == trump_suit for c in hand)
    ):
        return [c for c in hand if c.suit == trump_suit]
    return list(hand)


def _seats_holding_trump(
    sim_hands: dict[Seat, list[Card]], trump_suit: Suit | None
) -> set[Seat]:
    if trump_suit is None:
        return set()
    return {
        s for s, h in sim_hands.items()
        if any(c.suit == trump_suit for c in h)
    }


def _hand_remove(
    sim_hands: dict[Seat, list[Card]], seat: Seat, card: Card
) -> dict[Seat, list[Card]]:
    new = {s: list(cs) for s, cs in sim_hands.items()}
    new[seat].remove(card)
    return new


# ---------------------------------------------------------------------------
# Caps solver — boolean: caller's team wins every round.
# ---------------------------------------------------------------------------


def _solve_caps(
    *,
    sim_hands: dict[Seat, list[Card]],
    caller_seat: Seat,
    caller_order: list[Card],
    caller_index: int,
    leader: Seat,
    in_progress: list[tuple[Seat, Card]],
    rounds_remaining: int,
    trump_suit: Suit | None,
    my_team: Team,
    pcc_partner_out: Seat | None,
) -> bool:
    if rounds_remaining <= 0:
        return True

    turn_order = _round_turn_order(leader, pcc_partner_out)
    next_idx = len(in_progress)

    if next_idx >= len(turn_order):
        winner = _round_winner(in_progress, trump_suit)
        if team_of(winner) != my_team:
            return False
        if rounds_remaining == 1:
            return True
        return _solve_caps(
            sim_hands=sim_hands,
            caller_seat=caller_seat,
            caller_order=caller_order,
            caller_index=caller_index,
            leader=winner,
            in_progress=[],
            rounds_remaining=rounds_remaining - 1,
            trump_suit=trump_suit,
            my_team=my_team,
            pcc_partner_out=pcc_partner_out,
        )

    next_seat_to_play = turn_order[next_idx]
    led_suit: Suit | None = in_progress[0][1].suit if in_progress else None
    is_lead = not in_progress
    seats_with_trumps = _seats_holding_trump(sim_hands, trump_suit)

    if next_seat_to_play == caller_seat:
        if caller_index >= len(caller_order):
            return False
        card = caller_order[caller_index]
        hand = sim_hands.get(caller_seat, [])
        if card not in hand:
            return False
        legal = _legal_plays(
            hand=hand,
            led_suit=led_suit,
            trump_suit=trump_suit,
            is_lead=is_lead,
            seats_with_trumps=seats_with_trumps,
            seat=caller_seat,
        )
        if card not in legal:
            return False
        return _solve_caps(
            sim_hands=_hand_remove(sim_hands, caller_seat, card),
            caller_seat=caller_seat,
            caller_order=caller_order,
            caller_index=caller_index + 1,
            leader=leader,
            in_progress=in_progress + [(caller_seat, card)],
            rounds_remaining=rounds_remaining,
            trump_suit=trump_suit,
            my_team=my_team,
            pcc_partner_out=pcc_partner_out,
        )

    # Adversary: every legal choice must still preserve the sweep.
    other_hand = sim_hands.get(next_seat_to_play, [])
    if not other_hand:
        return False
    legal = _legal_plays(
        hand=other_hand,
        led_suit=led_suit,
        trump_suit=trump_suit,
        is_lead=is_lead,
        seats_with_trumps=seats_with_trumps,
        seat=next_seat_to_play,
    )
    if not legal:
        return False
    for chosen in legal:
        ok = _solve_caps(
            sim_hands=_hand_remove(sim_hands, next_seat_to_play, chosen),
            caller_seat=caller_seat,
            caller_order=caller_order,
            caller_index=caller_index,
            leader=leader,
            in_progress=in_progress + [(next_seat_to_play, chosen)],
            rounds_remaining=rounds_remaining,
            trump_suit=trump_suit,
            my_team=my_team,
            pcc_partner_out=pcc_partner_out,
        )
        if not ok:
            return False
    return True


# ---------------------------------------------------------------------------
# Claim-balance solver — integer: minimum points caller's team wins.
# ---------------------------------------------------------------------------


def _solve_min_points(
    *,
    sim_hands: dict[Seat, list[Card]],
    caller_seat: Seat,
    caller_order: list[Card],
    caller_index: int,
    leader: Seat,
    in_progress: list[tuple[Seat, Card]],
    rounds_remaining: int,
    trump_suit: Suit | None,
    my_team: Team,
    pcc_partner_out: Seat | None,
) -> int:
    if rounds_remaining <= 0:
        return 0

    turn_order = _round_turn_order(leader, pcc_partner_out)
    next_idx = len(in_progress)

    if next_idx >= len(turn_order):
        winner = _round_winner(in_progress, trump_suit)
        gained = _round_points(in_progress) if team_of(winner) == my_team else 0
        if rounds_remaining == 1:
            return gained
        return gained + _solve_min_points(
            sim_hands=sim_hands,
            caller_seat=caller_seat,
            caller_order=caller_order,
            caller_index=caller_index,
            leader=winner,
            in_progress=[],
            rounds_remaining=rounds_remaining - 1,
            trump_suit=trump_suit,
            my_team=my_team,
            pcc_partner_out=pcc_partner_out,
        )

    next_seat_to_play = turn_order[next_idx]
    led_suit: Suit | None = in_progress[0][1].suit if in_progress else None
    is_lead = not in_progress
    seats_with_trumps = _seats_holding_trump(sim_hands, trump_suit)

    if next_seat_to_play == caller_seat:
        if caller_index >= len(caller_order):
            return 0
        card = caller_order[caller_index]
        hand = sim_hands.get(caller_seat, [])
        if card not in hand:
            return 0
        legal = _legal_plays(
            hand=hand,
            led_suit=led_suit,
            trump_suit=trump_suit,
            is_lead=is_lead,
            seats_with_trumps=seats_with_trumps,
            seat=caller_seat,
        )
        if card not in legal:
            return 0
        return _solve_min_points(
            sim_hands=_hand_remove(sim_hands, caller_seat, card),
            caller_seat=caller_seat,
            caller_order=caller_order,
            caller_index=caller_index + 1,
            leader=leader,
            in_progress=in_progress + [(caller_seat, card)],
            rounds_remaining=rounds_remaining,
            trump_suit=trump_suit,
            my_team=my_team,
            pcc_partner_out=pcc_partner_out,
        )

    other_hand = sim_hands.get(next_seat_to_play, [])
    if not other_hand:
        return 0
    legal = _legal_plays(
        hand=other_hand,
        led_suit=led_suit,
        trump_suit=trump_suit,
        is_lead=is_lead,
        seats_with_trumps=seats_with_trumps,
        seat=next_seat_to_play,
    )
    if not legal:
        return 0
    # Adversary minimises caller's-team points.
    best: int | None = None
    for chosen in legal:
        v = _solve_min_points(
            sim_hands=_hand_remove(sim_hands, next_seat_to_play, chosen),
            caller_seat=caller_seat,
            caller_order=caller_order,
            caller_index=caller_index,
            leader=leader,
            in_progress=in_progress + [(next_seat_to_play, chosen)],
            rounds_remaining=rounds_remaining,
            trump_suit=trump_suit,
            my_team=my_team,
            pcc_partner_out=pcc_partner_out,
        )
        if best is None or v < best:
            best = v
    return best if best is not None else 0

File: engine/dd.ts
// Per-world double-dummy solver. Given a fixed world W (concrete card
// locations) and a fixed caller play order O, decide whether O wins
// every remaining round (or, for claim balance, the minimum points the
// caller's team is guaranteed to collect) against any legal opponent
// play. Mirrors game304/dd.py.
//
// The outer single-dummy quantifiers (universal over consistent worlds,
// existential over orders) live in caps.ts. This module is a leaf:
// it knows nothing about information sets.

import type { CardId, Suit } from './card';
import { suitOf } from './card';
import { legalPlays, roundPoints, roundTurnOrder, roundWinner, seatsHoldingTrump } from './play';
import type { Seat, Team } from './seating';
import { SEAT_INDEX, SEATS_BY_INDEX, teamOf } from './seating';
import type { World } from './info';

export interface InProgressEntry {
  seat: Seat;
  card: CardId;
}

export interface PlaySnapshot {
  leader: Seat;
  entries: ReadonlyArray<InProgressEntry>;
}

export interface OrderSweepsArgs {
  world: World;
  callerSeat: Seat;
  callerOrder: ReadonlyArray<CardId>;
  snapshot: PlaySnapshot;
  pccPartnerOut: Seat | null;
  roundsRemaining: number;
}

const worldHandsToMap = (
  hands: ReadonlyArray<ReadonlyArray<CardId>>,
): Map<Seat, CardId[]> => {
  const m = new Map<Seat, CardId[]>();
  for (let i = 0; i < 4; i++) {
    m.set(SEATS_BY_INDEX[i], [...(hands[i] ?? [])]);
  }
  return m;
};

const mapToHandsArr = (
  hands: ReadonlyMap<Seat, ReadonlyArray<CardId>>,
): ReadonlyArray<ReadonlyArray<CardId>> => {
  const arr: ReadonlyArray<CardId>[] = [[], [], [], []];
  for (const [s, cs] of hands) arr[SEAT_INDEX[s]] = cs;
  return arr;
};


// Equivalence-class reduction for a solver's move list.
//
// Two cards of the same suit in the SAME hand are strategically
// identical when no card of that suit still in play sits between them:
// whichever you play, the other takes its place in every subsequent
// position. Collapsing those to one representative is the standard
// double-dummy pruning, and without it these searches are unusable —
// a seven-round position branches roughly (7^3)^7 before pruning.
//
// `inPlay` must be every card not yet played: all four hands plus the
// cards already on the table this trick.
const reduceEquivalent = (
  legal: ReadonlyArray<CardId>,
  hand: ReadonlyArray<CardId>,
  inPlay: ReadonlySet<CardId>,
): CardId[] => {
  if (legal.length <= 1) return [...legal];
  const RANKS = ['J', '9', 'A', '10', 'K', 'Q', '8', '7'];
  const rank = (c: CardId): number => RANKS.indexOf(c.slice(0, c.length - 1));
  const own = new Set(hand);
  const keep: CardId[] = [];
  const bySuit = new Map<Suit, CardId[]>();
  for (const c of legal) {
    const su = suitOf(c);
    if (!bySuit.has(su)) bySuit.set(su, []);
    bySuit.get(su)!.push(c);
  }
  for (const [su, cards] of bySuit) {
    // All still-in-play cards of this suit, strongest first.
    const global = [...inPlay].filter(c => suitOf(c) === su).sort((a, b) => rank(a) - rank(b));
    const candidate = new Set(cards);
    let prevWasOursAndKept = false;
    for (const c of global) {
      if (!own.has(c)) { prevWasOursAndKept = false; continue; }
      if (!candidate.has(c)) { prevWasOursAndKept = false; continue; }
      // Consecutive in the global ordering and both ours ⇒ equivalent.
      if (!prevWasOursAndKept) keep.push(c);
      prevWasOursAndKept = true;
    }
  }
  return keep.length > 0 ? keep : [...legal];
};

export const orderSweepsWorld = (args: OrderSweepsArgs): boolean => {
  const simHands = worldHandsToMap(args.world.hands);
  const inProgress: Array<[Seat, CardId]> =
    args.snapshot.entries.map(e => [e.seat, e.card]);
  return solveCaps({
    simHands,
    callerSeat: args.callerSeat,
    callerOrder: [...args.callerOrder],
    callerIndex: 0,
    leader: args.snapshot.leader,
    inProgress,
    roundsRemaining: args.roundsRemaining,
    trumpSuit: args.world.trumpSuit,
    myTeam: teamOf(args.callerSeat),
    pccPartnerOut: args.pccPartnerOut,
  });
};

export const orderMinPointsInWorld = (args: OrderSweepsArgs): number => {
  const simHands = worldHandsToMap(args.world.hands);
  const inProgress: Array<[Seat, CardId]> =
    args.snapshot.entries.map(e => [e.seat, e.card]);
  return solveMinPoints({
    simHands,
    callerSeat: args.callerSeat,
    callerOrder: [...args.callerOrder],
    callerIndex: 0,
    leader: args.snapshot.leader,
    inProgress,
    roundsRemaining: args.roundsRemaining,
    trumpSuit: args.world.trumpSuit,
    myTeam: teamOf(args.callerSeat),
    pccPartnerOut: args.pccPartnerOut,
  });
};

interface SolveCtx {
  simHands: Map<Seat, CardId[]>;
  callerSeat: Seat;
  callerOrder: CardId[];
  callerIndex: number;
  leader: Seat;
  inProgress: Array<[Seat, CardId]>;
  roundsRemaining: number;
  trumpSuit: Suit;
  myTeam: Team;
  pccPartnerOut: Seat | null;
}

const handRemove = (
  hands: Map<Seat, CardId[]>,
  seat: Seat,
  card: CardId,
): Map<Seat, CardId[]> => {
  const next = new Map<Seat, CardId[]>();
  for (const [s, cs] of hands) {
    if (s === seat) {
      const idx = cs.indexOf(card);
      if (idx === -1) {
        next.set(s, [...cs]);
      } else {
        const copy = [...cs];
        copy.splice(idx, 1);
        next.set(s, copy);
      }
    } else {
      next.set(s, [...cs]);
    }
  }
  return next;
};

const solveCaps = (ctx: SolveCtx): boolean => {
  if (ctx.roundsRemaining <= 0) return true;

  const turnOrder = roundTurnOrder(ctx.leader, ctx.pccPartnerOut);
  const nextIdx = ctx.inProgress.length;

  if (nextIdx >= turnOrder.length) {
    const winner = roundWinner(ctx.inProgress, ctx.trumpSuit);
    if (teamOf(winner) !== ctx.myTeam) return false;
    if (ctx.roundsRemaining === 1) return true;
    return solveCaps({
      ...ctx,
      leader: winner,
      inProgress: [],
      roundsRemaining: ctx.roundsRemaining - 1,
    });
  }

  const nextSeatToPlay = turnOrder[nextIdx];
  const ledSuit: Suit | null =
    ctx.inProgress.length > 0 ? suitOf(ctx.inProgress[0][1]) : null;
  const isLead = ctx.inProgress.length === 0;
  const trumpHolders = seatsHoldingTrump(mapToHandsArr(ctx.simHands), ctx.trumpSuit);

  if (nextSeatToPlay === ctx.callerSeat) {
    if (ctx.callerIndex >= ctx.callerOrder.length) return false;
    const card = ctx.callerOrder[ctx.callerIndex];
    const hand = ctx.simHands.get(ctx.callerSeat) ?? [];
    if (!hand.includes(card)) return false;
    const legal = legalPlays({
      hand,
      ledSuit,
      trumpSuit: ctx.trumpSuit,
      isLead,
      seatsWithTrumps: trumpHolders,
      seat: ctx.callerSeat,
    });
    if (!legal.includes(card)) return false;
    return solveCaps({
      ...ctx,
      simHands: handRemove(ctx.simHands, ctx.callerSeat, card),
      callerIndex: ctx.callerIndex + 1,
      inProgress: [...ctx.inProgress, [ctx.callerSeat, card]],
    });
  }

  const otherHand = ctx.simHands.get(nextSeatToPlay) ?? [];
  if (otherHand.length === 0) return false;
  const legal = legalPlays({
    hand: otherHand,
    ledSuit,
    trumpSuit: ctx.trumpSuit,
    isLead,
    seatsWithTrumps: trumpHolders,
    seat: nextSeatToPlay,
  });
  if (legal.length === 0) return false;

  const inPlay = new Set<CardId>();
  for (const [, cs] of ctx.simHands) for (const c of cs) inPlay.add(c);
  for (const [, c] of ctx.inProgress) inPlay.add(c);
  for (const chosen of reduceEquivalent(legal, otherHand, inPlay)) {
    const ok = solveCaps({
      ...ctx,
      simHands: handRemove(ctx.simHands, nextSeatToPlay, chosen),
      inProgress: [...ctx.inProgress, [nextSeatToPlay, chosen]],
    });
    if (!ok) return false;
  }
  return true;
};

const solveMinPoints = (ctx: SolveCtx): number => {
  if (ctx.roundsRemaining <= 0) return 0;

  const turnOrder = roundTurnOrder(ctx.leader, ctx.pccPartnerOut);
  const nextIdx = ctx.inProgress.length;

  if (nextIdx >= turnOrder.length) {
    const winner = roundWinner(ctx.inProgress, ctx.trumpSuit);
    const gained =
      teamOf(winner) === ctx.myTeam ? roundPoints(ctx.inProgress) : 0;
    if (ctx.roundsRemaining === 1) return gained;
    return gained + solveMinPoints({
      ...ctx,
      leader: winner,
      inProgress: [],
      roundsRemaining: ctx.roundsRemaining - 1,
    });
  }

  const nextSeatToPlay = turnOrder[nextIdx];
  const ledSuit: Suit | null =
    ctx.inProgress.length > 0 ? suitOf(ctx.inProgress[0][1]) : null;
  const isLead = ctx.inProgress.length === 0;
  const trumpHolders = seatsHoldingTrump(mapToHandsArr(ctx.simHands), ctx.trumpSuit);

  if (nextSeatToPlay === ctx.callerSeat) {
    if (ctx.callerIndex >= ctx.callerOrder.length) return 0;
    const card = ctx.callerOrder[ctx.callerIndex];
    const hand = ctx.simHands.get(ctx.callerSeat) ?? [];
    if (!hand.includes(card)) return 0;
    const legal = legalPlays({
      hand,
      ledSuit,
      trumpSuit: ctx.trumpSuit,
      isLead,
      seatsWithTrumps: trumpHolders,
      seat: ctx.callerSeat,
    });
    if (!legal.includes(card)) return 0;
    return solveMinPoints({
      ...ctx,
      simHands: handRemove(ctx.simHands, ctx.callerSeat, card),
      callerIndex: ctx.callerIndex + 1,
      inProgress: [...ctx.inProgress, [ctx.callerSeat, card]],
    });
  }

  const otherHand = ctx.simHands.get(nextSeatToPlay) ?? [];
  if (otherHand.length === 0) return 0;
  const legal = legalPlays({
    hand: otherHand,
    ledSuit,
    trumpSuit: ctx.trumpSuit,
    isLead,
    seatsWithTrumps: trumpHolders,
    seat: nextSeatToPlay,
  });
  if (legal.length === 0) return 0;

  let best: number | null = null;
  for (const chosen of legal) {
    const v = solveMinPoints({
      ...ctx,
      simHands: handRemove(ctx.simHands, nextSeatToPlay, chosen),
      inProgress: [...ctx.inProgress, [nextSeatToPlay, chosen]],
    });
    if (best === null || v < best) best = v;
  }
  return best ?? 0;
};

// Adaptive per-world sweep — the honest reading of `∃σ_W` from
// [docs/specs/caps_formalism.md §5], as opposed to `orderSweepsWorld`'s
// fixed-order `∃O`. Within a fully-known world the caller chooses each
// card knowing what has been played, and every other seat — partner
// included, per §5's "V still cannot rely on partner choice" — plays
// adversarially.
//
// Used to verify the CSP's obligation predicate independently: if the
// CSP says obligated but some consistent world fails this, the CSP is
// unsound. See tools/puzzles/audit-obligation.ts.
export interface WorldSweepArgs {
  hands: ReadonlyMap<Seat, ReadonlyArray<CardId>>;
  callerSeat: Seat;
  leader: Seat;
  inProgress: ReadonlyArray<readonly [Seat, CardId]>;
  roundsRemaining: number;
  trumpSuit: Suit;
  pccPartnerOut?: Seat | null;
}

export const worldSweepsAdaptive = (args: WorldSweepArgs): boolean => {
  const pccOut = args.pccPartnerOut ?? null;
  const rec = (
    hands: Map<Seat, CardId[]>,
    leader: Seat,
    inProgress: Array<readonly [Seat, CardId]>,
    roundsRemaining: number,
  ): boolean => {
    if (roundsRemaining <= 0) return true;
    const order = roundTurnOrder(leader, pccOut);
    if (inProgress.length >= order.length) {
      const winner = roundWinner(inProgress, args.trumpSuit);
      if (teamOf(winner) !== teamOf(args.callerSeat)) return false;
      if (roundsRemaining === 1) return true;
      return rec(hands, winner, [], roundsRemaining - 1);
    }
    const seat = order[inProgress.length];
    const hand = hands.get(seat) ?? [];
    if (hand.length === 0) return false;
    const arr: ReadonlyArray<CardId>[] = [[], [], [], []];
    for (let i = 0; i < 4; i++) arr[i] = hands.get(SEATS_BY_INDEX[i]) ?? [];
    const legal = legalPlays({
      hand,
      ledSuit: inProgress.length > 0 ? suitOf(inProgress[0][1]) : null,
      trumpSuit: args.trumpSuit,
      isLead: inProgress.length === 0,
      seatsWithTrumps: seatsHoldingTrump(arr, args.trumpSuit),
      seat,
    });
    if (legal.length === 0) return false;
    const isCaller = seat === args.callerSeat;
    const inPlay = new Set<CardId>();
    for (const [, cs] of hands) for (const c of cs) inPlay.add(c);
    for (const [, c] of inProgress) inPlay.add(c);
    for (const c of reduceEquivalent(legal, hand, inPlay)) {
      const next = new Map(hands);
      next.set(seat, hand.filter(x => x !== c));
      const ok = rec(next, leader, [...inProgress, [seat, c] as const], roundsRemaining);
      if (isCaller) { if (ok) return true; }
      else if (!ok) return false;
    }
    return !isCaller;
  };
  const start = new Map<Seat, CardId[]>();
  for (const [s, cs] of args.hands) start.set(s, [...cs]);
  return rec(start, args.leader, [...args.inProgress], args.roundsRemaining);
};
